src/class/cdc/cdc_host.c

```diff
@@ -1171,4 +1170,351 @@ static void cp210x_process_config(tuh_xfer_t* xfer) {
+    control |= CH34X_BIT_DTR;
+  }
+  TU_ASSERT (ch34x_control_out(p_cdc, CH34X_REQ_MODEM_CTRL, (uint8_t) ~control, 0,
+                               complete_cb ? ch34x_control_complete : NULL, user_data));
+  return true;
+}
+
+static void ch34x_process_config(tuh_xfer_t* xfer) {
+  uint8_t const itf_num = ch34x_xfer_get_itf_num(xfer);
+  uintptr_t const state = xfer->user_data;
+  uint8_t const idx = tuh_cdc_itf_get_index(xfer->daddr, itf_num);
+  cdch_interface_t* p_cdc = get_itf(idx);
+  cdc_line_coding_t const line_coding = CFG_TUH_CDC_LINE_CODING_ON_ENUM_CH34X;
+  uint8_t buffer[2];
+  TU_ASSERT (p_cdc,);
+
+  switch (state) {
+    case CONFIG_CH34X_READ_VERSION: // request version read
+      #if CH34X_LOGS
+        TU_LOG_DRV ( "[%u] CDCh CH34x Process Config started\r\n", p_cdc->daddr );
+      #endif
+      TU_ASSERT (ch34x_control_in(p_cdc, CH34X_REQ_READ_VERSION, 0, 0, buffer, 2, ch34x_process_config, CONFIG_CH34X_SERIAL_INIT),);
+      break;
+
+    case CONFIG_CH34X_SERIAL_INIT: { // handle version read data, request to init CH34x with line_coding and baudrate
+      uint8_t version = xfer->buffer[0];
+      #if CH34X_LOGS
+        TU_LOG_DRV("[%u] CDCh CH34x Chip version=%02x\r\n", p_cdc->daddr, version);
+      #endif
+      // only versions >= 0x30 are tested, below 0x30 seems having other programming, see WCH vendor, linux kernel and FreeBSD drivers
+      TU_ASSERT (version >= 0x30,);
+      uint8_t factor, divisor;
+      TU_ASSERT (ch34x_get_factor_divisor(line_coding.bit_rate, &factor, &divisor),);
+      uint8_t lcr = CH34X_LCR_ENABLE_RX | CH34X_LCR_ENABLE_TX;
+      switch (line_coding.data_bits) {
+        case 5:
+          lcr |= CH34X_LCR_CS5;
+          break;
+        case 6:
+          lcr |= CH34X_LCR_CS6;
+          break;
+        case 7:
+          lcr |= CH34X_LCR_CS7;
+          break;
+        case 8:
+          lcr |= CH34X_LCR_CS8;
+          break;
+        default:
+          TU_ASSERT (false,); // not supported data_bits
+          lcr |= CH34X_LCR_CS8;
+          break;
+      }
+
+      if (line_coding.parity != CDC_LINE_CODING_PARITY_NONE) {
+        lcr |= CH34X_LCR_ENABLE_PAR;
+        if (line_coding.parity == CDC_LINE_CODING_PARITY_EVEN ||
+            line_coding.parity == CDC_LINE_CODING_PARITY_SPACE) {
+          lcr |= CH34X_LCR_PAR_EVEN;
+        }
+        if (line_coding.parity == CDC_LINE_CODING_PARITY_MARK ||
+            line_coding.parity == CDC_LINE_CODING_PARITY_SPACE) {
+          lcr |= CH34X_LCR_MARK_SPACE;
+        }
+      }
+      TU_ASSERT (line_coding.stop_bits == CDC_LINE_CODING_STOP_BITS_1 ||
+                 line_coding.stop_bits == CDC_LINE_CODING_STOP_BITS_2,); // 1.5 stop bits not supported
+      if (line_coding.stop_bits == CDC_LINE_CODING_STOP_BITS_2) {
+        lcr |= CH34X_LCR_STOP_BITS_2;
+      }
+      TU_ASSERT (ch34x_control_out(p_cdc, CH34X_REQ_SERIAL_INIT, (uint16_t) (lcr << 8 | 0x9c), (uint16_t) (factor << 8 | 0x80 | divisor),
```

Comment:
both https://github.com/OLIMEX/ch340-dkms/blob/master/ch340.c#L282 and linux run this commad with (0,0) then set baudrate and lsr afterwards. Should we follow their init sequence as well ?

Reply:
My list of drivers examined above was not complete.
As already mentioned, I examined the Linux sources of:
- Vendor WCH (with poor quality)
- Linux Kernel
- FreeBSD

None of the 3 Linux drivers worked with all of my CH34x on the desk, not even the WCH one!
Then I tried the two different WCH Windows driver
- V3.5.2019.1
- V3.8.2023.2

Only V3.5.2019.1 worked with all my CH34x.
I analyzed both versions with an USB bus analyzer, get the working init and config sequence and overtaken them in my code.
To be clear, the original WCH manufacturer driver for Linux does not work and behaves differently than the original WCH manufacturer driver for Windows.
I may have a clone and not an original WCH CH34x on my desk and WCH may have changed its driver in the meantime so that it no longer works with the clones.
Quite ugly...

To your question:
The fully functioning WCH Windows driver uses the CH34X_REQ_SERIAL_INIT (CMD_C1 in WCH driver) request to configure line coding and baud rate (factor and divisor) at the same time with one control request.
The WCH Linux driver uses the sequence CH34X_REQ_SERIAL_INIT (0,0) and then writing the registers for line coding and baudrate, but as I said, this is also possible together in a CH34X_REQ_SERIAL_INIT.
CH34X_REQ_SERIAL_INIT may perform additional initializations...

After this initial configuration, the driver sends register write requests to the line coding register LCR (0x18) and factor/divisor registers (0x12 and 0x13) if corresponding settings are to be changed during operation.
Some CH34x devices before version 0x30 may also use LCR2 register (0x25), but I've no device on my desk to verify it.
See https://github.com/torvalds/linux/blob/052d534373b7ed33712a63d5e17b2b6cdbce84fd/drivers/usb/serial/ch341.c#L272

Two registers can be written simultaneously as part of a register write request.
wValue contains the addresses of the two registers in low and high bytes and wIndex contains the register values to be written also in low and high bytes.
See https://github.com/freebsd/freebsd-src/blob/a079c891c01b1f8b32d79b490d9e7b7542e26d0f/sys/dev/usb/serial/uchcom.c#L446

Reply:
thank you for clarifcation, I also take the peek at linux and wch linux driver. Didn't know there is an windows source driver that init with these 2 extra parameters.